File: agent/agent.py
"""
agent.py — the local BombosSwap player agent.

Sits next to one player's emulator (EmuNWA/RetroArch) or FXPak/SD2SNES and:
  * polls WRAM to detect items the player picks up in their own world,
  * reports those pickups to the coordination server,
  * applies the server's grant/revoke commands by writing WRAM,
  * dials OUT to the server over WebSocket (NAT-friendly remote play).

Echo-cancellation: every write we make is recorded as an "expected" byte at that
address so our own poll loop doesn't mistake a grant/revoke for a fresh pickup.
"""

import json
import logging
import os
import threading
import time

import websocket  # websocket-client

from shared.items import ITEMS, BY_KEY, discovered_level
from shared import protocol as P
from .effects import Effects
from .sni.hud_text import HudText
from .sni.memory_constants import MEMORY_ADDRESSES, OUT_OF_GAME_MODES, PLAYABLE_MODES
from .sni.item_effects import ABILITY_ADDR, RUN_ABILITY_MASK

logger = logging.getLogger("HyruleAgent")

# BillognaBot's HUD text endpoint (see HyruleAgent._hud_show). Empty = never
# forward, always draw the strip ourselves.
BOT_HUD_URL = (os.environ.get("HYRULELINK_BOT_HUD", "http://127.0.0.1:5000/api/hud/say")
               if os.environ.get("HYRULELINK_BOT_HUD", "1") not in ("0", "", "off", "no")
               else "")
if BOT_HUD_URL and not BOT_HUD_URL.startswith("http"):
    BOT_HUD_URL = "http://127.0.0.1:5000/api/hud/say"
BOT_HUD_TIMEOUT_S = 1.5     # a bot mid-startup can hold the socket this long
BOT_HUD_RETRY_S = 30.0      # after a refused/failed call, draw locally this long

GAME_MODE_ADDR = MEMORY_ADDRESSES["game_mode"]  # 0x0010

# Items grouped by the WRAM address they live in (bitfields share an address).
_ITEMS_BY_ADDR = {}
for _it in ITEMS:
    _ITEMS_BY_ADDR.setdefault(_it.addr, []).append(_it)
_TRACKED_ADDRS = sorted(_ITEMS_BY_ADDR.keys())
_TRACKED_START = _TRACKED_ADDRS[0]
_TRACKED_SIZE = _TRACKED_ADDRS[-1] - _TRACKED_START + 1


class HyruleAgent:
    def __init__(self, transport, server_ws_url, room, user_id, player_token,
                 poll_interval=1.0, on_notify=None, hud_text=True, bot_hud_url=None):
        self.t = transport
        self.fx = Effects(transport)
        self.url = server_ws_url
        self.room = room
        self.user_id = int(user_id)
        self.token = player_token
        self.poll_interval = poll_interval
        # optional callback(text) so a host app (the desktop GUI) can surface
        # server notifications even when the transport has no OSD (Snes9x-NWA,
        # SNI/hardware — only RetroArch can draw on-screen messages itself).
        self.on_notify = on_notify
        # true in-game messages: rendered on the HUD strip via WRAM writes
        # (works on every transport, no ROM patch — see sni/hud_text.py)
        self.hud = HudText(transport) if hud_text else None
        # BillognaBot's /api/hud/say, or "" to always draw the strip ourselves
        # (see _hud_show). Callers pass BOT_HUD_URL; tests leave it off.
        self._bot_hud_url = bot_hud_url or ""
        # when the bot's endpoint last failed (0 = try it): _hud_show
        self._bot_hud_retry_at = 0.0

        self.ws = None
        self._ws_ready = threading.Event()
        self._stop = threading.Event()

        # echo-cancellation + change detection (shared between ws + poll threads)
        self._lock = threading.Lock()
        self._baseline = {}   # addr -> last raw byte we accept as "known"
        self._expected = {}   # addr -> raw byte we just wrote (suppress once)
        # Grants/revokes that arrived while no save was loaded (title screen /
        # file select). Writing then would land in a stale $7EF000 mirror and be
        # wiped by the next file load, so they wait here until gameplay resumes.
        self._pending = {}    # item key -> (level, enable)
        # True after MODE passes through an out-of-game module (file select,
        # save-quit, loading). Re-entering gameplay from there means the WRAM
        # mirror may be a different/re-loaded save: re-seed and resync.
        self._out_of_game = False
        # Whether we currently own the boots. ALttP can clear the dash-ability
        # flag ($7EF379 bit 0x04) on screen transitions, so we re-assert it each
        # poll while owned — otherwise "have boots but can't run".
        self._boots_owned = False
        # Track emulator connectivity so we can re-sync after a crash/reload.
        self._emu_was_connected = False

    # ── WebSocket plumbing ─────────────────────────────────────────────────
    def _on_open(self, ws):
        ws.send(json.dumps({
            "type": P.HELLO, "role": P.ROLE_AGENT,
            "room": self.room, "player_id": self.user_id, "token": self.token,
        }))
        self._ws_ready.set()
        self._send_status()  # report current emulator connectivity
        logger.info("Connected to server; agent online.")

    def _send_status(self):
        """Tell the server whether our emulator is currently reachable."""
        if self.ws and self._ws_ready.is_set():
            try:
                self.ws.send(json.dumps({"type": P.STATUS, "emu": self.t.connected}))
            except Exception as e:
                logger.debug("status send failed: %s", e)

    def _on_message(self, ws, message):
        try:
            msg = json.loads(message)
        except Exception:
            return
        mtype = msg.get("type")
        if mtype == P.GRANT:
            self._apply(msg["item"], int(msg.get("level", 1)), enable=True)
        elif mtype == P.REVOKE:
            self._apply(msg["item"], 0, enable=False)
        elif mtype == P.NOTIFY:
            self._show_notification(msg.get("text", ""))
        elif mtype == P.REJECT:
            logger.info("Server rejected: %s", msg.get("reason"))

    def _show_notification(self, text):
        text = " ".join(str(text).split())[:120]
        if not text:
            return
        logger.info("• %s", text)          # always visible in the console/app log
        if self.on_notify:
            try:
                self.on_notify(text)       # desktop-app toast (works on any transport)
            except Exception as e:
                logger.debug("notify callback failed: %s", e)
        if self.hud:
            self._hud_show(text)           # in-game HUD strip (drawn by the poll loop)
        show = getattr(self.t, "show_message", None)
        if show and self.t.connected:
            try:
                show(text)                 # emulator OSD (RetroArch only)
            except Exception as e:
                logger.debug("emulator notification failed: %s", e)

    # ----- in-game HUD strip: one writer at a time -------------------------
    # BillognaBot (the streaming bot) writes chat/bits/points/welcome lines to
    # the same 20-cell HUD strip. When it is running on this PC it owns the
    # strip, so our messages go through its queue (POST /api/hud/say) instead
    # of racing it cell-for-cell; when it is not running, or it says it can't
    # draw right now (emulator not connected there), we draw ourselves.
    # Disable with HYRULELINK_BOT_HUD=0 (or a blank URL).

    def _hud_show(self, text):
        if not self._bot_hud_url or self._bot_hud_retry_at > time.time():
            self.hud.show(text)
            return
        # Off the websocket thread: a bot that is starting up can hold the
        # connection for the full timeout and we must not stall grant/revoke.
        threading.Thread(target=self._hud_forward, args=(text,), daemon=True).start()

    def _hud_forward(self, text):
        import urllib.request
        body = json.dumps({"text": text, "source": "hyrulelink"}).encode("utf-8")
        req = urllib.request.Request(self._bot_hud_url, data=body, method="POST",
                                     headers={"Content-Type": "application/json"})
        try:
            with urllib.request.urlopen(req, timeout=BOT_HUD_TIMEOUT_S) as resp:
                reply = json.loads(resp.read().decode("utf-8") or "{}")
            if reply.get("ok") and reply.get("shown"):
                return                     # the bot drew it
            logger.debug("bot HUD declined (%s); drawing locally", reply)
        except Exception as e:
            # Not running (connection refused), or not ready: draw locally and
            # don't knock again for a while.
            self._bot_hud_retry_at = time.time() + BOT_HUD_RETRY_S
            logger.debug("bot HUD unreachable (%s); drawing locally", e)
        self.hud.show(text)

    def _on_error(self, ws, err):
        logger.debug("ws error: %s", err)

    def _on_close(self, ws, *a):
        self._ws_ready.clear()
        logger.info("Server connection closed.")

    def _run_ws(self):
        while not self._stop.is_set():
            try:
                self.ws = websocket.WebSocketApp(
                    self.url,
                    on_open=self._on_open,
                    on_message=self._on_message,
                    on_error=self._on_error,
                    on_close=self._on_close,
                )
                self.ws.run_forever(ping_interval=30, ping_timeout=10)
            except Exception as e:
                logger.debug("ws loop error: %s", e)
            if not self._stop.is_set():
                time.sleep(3)  # reconnect backoff

    # ── applying server commands ───────────────────────────────────────────
    def _apply(self, key, level, enable, tries=3):
        if key not in BY_KEY:
            return
        # Only write while a save is actually loaded. Outside a playable module
        # (title screen, file select, loading) the $7EF000 mirror isn't the live
        # save — a write there is wiped by the next file load, leaving the game
        # out of step with the ledger. Defer instead; the poll loop flushes the
        # queue the moment gameplay resumes. (An unreadable mode byte is treated
        # as a transient hiccup and handled by the retry loop below.)
        gm = self.t.read_memory(GAME_MODE_ADDR, size=1)
        if gm and int(gm[0]) not in PLAYABLE_MODES:
            with self._lock:
                self._pending[key] = (level, enable)
            logger.info("Deferred %s of %s — no save loaded (mode 0x%02X)",
                        "grant" if enable else "revoke", key, int(gm[0]))
            return
        with self._lock:
            self._pending.pop(key, None)   # this command supersedes any deferred one
        # Grants/revokes are idempotent (they set absolute state), so retry a few
        # times across a transient emulator hiccup rather than silently dropping
        # the item — a dropped write/read must not leave the player without an item
        # the ledger says they own.
        last = None
        for _ in range(tries):
            if not self.t.connected:
                self.t.connect()
            try:
                if enable:
                    raw = self.fx.enable(key, level)
                    logger.info("Granted %s (lvl %s)", key, level)
                else:
                    raw = self.fx.disable(key)
                    logger.info("Revoked %s", key)
                if key == "boots":
                    self._boots_owned = enable
                # Suppress our own write so the poller doesn't re-broadcast it.
                with self._lock:
                    self._expected[BY_KEY[key].addr] = raw
                self._send_applied(key, enable, True)
                return
            except Exception as e:
                last = e
        logger.warning("apply %s failed after %d tries: %s", key, tries, last)
        self._send_applied(key, enable, False, str(last))

    def _send_applied(self, key, enable, ok, error=None):
        if not (self.ws and self._ws_ready.is_set()):
            return
        payload = {"type": P.APPLIED, "item": key,
                   "action": "grant" if enable else "revoke", "ok": bool(ok)}
        if error:
            payload["error"] = error[:200]
        try:
            self.ws.send(json.dumps(payload))
        except Exception as e:
            logger.debug("apply acknowledgement failed: %s", e)

    # ── pickup detection poll loop ─────────────────────────────────────────
    def _send_pickup(self, key, level):
        if self.ws and self._ws_ready.is_set():
            try:
                self.ws.send(json.dumps({"type": P.PICKUP, "item": key, "level": level}))
                logger.info("Reported pickup: %s lvl %s", key, level)
            except Exception as e:
                logger.debug("send pickup failed: %s", e)

    def _poll_once(self):
        # Only trust the SRAM mirror while actually in a playable game mode;
        # at file-select / transitions the bytes can be stale or zeroed.
        gm = self.t.read_memory(GAME_MODE_ADDR, size=1)
        if not gm or int(gm[0]) not in PLAYABLE_MODES:
            # Passing through the title/file-select/loading modules means the
            # next playable state may be a different or re-loaded save.
            if gm and int(gm[0]) in OUT_OF_GAME_MODES:
                self._out_of_game = True
            return
        if self._out_of_game:
            self._out_of_game = False
            # A (re)loaded save reverts WRAM to what was last saved: items we
            # revoked reappear and items we granted vanish. Re-seed detection so
            # the reverted bytes aren't reported as fresh pickups, and have the
            # server re-push ownership so the game matches the ledger again.
            self._resync("Save (re)loaded")
        self._flush_pending()
        if self.hud:
            self.hud.tick()   # only while playable — the HUD buffer is live here

        # All inventory bytes occupy one small contiguous SRAM-mirror range.
        # Reading it once cuts RetroArch UDP traffic from ~23 round trips per
        # poll to one, which materially reduces dropped replies and UI flicker.
        snapshot = self.t.read_memory(_TRACKED_START, size=_TRACKED_SIZE)
        if not snapshot or len(snapshot) < _TRACKED_SIZE:
            return

        for addr in _TRACKED_ADDRS:
            raw = int(snapshot[addr - _TRACKED_START])
            with self._lock:
                old = self._baseline.get(addr)
                exp = self._expected.get(addr)
                if old is None:
                    self._baseline[addr] = raw      # first sight, seed baseline
                    continue
                if exp is not None and raw == exp:
                    self._baseline[addr] = raw       # our own write echoing back
                    del self._expected[addr]
                    continue
                if raw == old:
                    continue
                pickups = []
                for item in _ITEMS_BY_ADDR[addr]:
                    nl = discovered_level(item, raw)
                    ol = discovered_level(item, old)
                    if nl > ol:                      # 0->1, or a progressive bump
                        pickups.append((item.key, nl))
                self._baseline[addr] = raw
            for key, level in pickups:
                self._send_pickup(key, level)

        ability = (int(snapshot[ABILITY_ADDR - _TRACKED_START])
                   if _TRACKED_START <= ABILITY_ADDR < _TRACKED_START + _TRACKED_SIZE else None)
        self._enforce_boots_ability(ability)

    def _enforce_boots_ability(self, ability_byte=None):
        """Keep the dash-ability flag ($7EF379 bit 0x04) in sync with ownership.

        Dashing is gated by this flag, not the boots inventory byte, and ALttP
        can both clear it on transitions (own boots but can't run) and leave it
        set after a lost revoke (run without boots). So every poll: set it if we
        own the boots, clear it if we don't. Writes only when it's actually
        wrong, so it's cheap and self-heals dropped grant/revoke writes."""
        if ability_byte is None:
            data = self.t.read_memory(ABILITY_ADDR, size=1)
            if not data:
                return
            ability_byte = int(data[0])
        have = bool(ability_byte & RUN_ABILITY_MASK)
        if self._boots_owned and not have:
            self.t.write_memory(ABILITY_ADDR, bytes([ability_byte | RUN_ABILITY_MASK]))
        elif not self._boots_owned and have:
            self.t.write_memory(ABILITY_ADDR, bytes([ability_byte & ~RUN_ABILITY_MASK]))

    def _flush_pending(self):
        """Apply grants/revokes that were deferred while no save was loaded."""
        with self._lock:
            pending = list(self._pending.items())
            self._pending.clear()
        for key, (level, enable) in pending:
            self._apply(key, level, enable)

    def _resync(self, reason, wipe_baseline=True):
        """Ask the server to re-push our ownership so the game is reconciled to
        the authoritative ledger. Used when the emulator comes back (crash/
        reload) and when a save is (re)loaded.

        `wipe_baseline` re-seeds pickup detection from current WRAM WITHOUT
        reporting its contents as fresh pickups — correct when the save may
        genuinely have changed underneath us (reload/restart), since a stale
        mirror shouldn't be reported as a find. But a plain transport blip (the
        emulator's command port hiccuped; the same save kept running the whole
        time) must NOT wipe it: the old baseline is still valid, and doing so
        would silently swallow any pickup that happened during the outage —
        the server's next ownership push then has no record of it and revokes
        an item the player legitimately holds."""
        logger.info("%s — %s.", reason,
                    "re-seeding state and requesting resync" if wipe_baseline
                    else "requesting resync")
        if wipe_baseline:
            with self._lock:
                self._baseline.clear()
                self._expected.clear()
            if self.hud:
                self.hud.reset()   # the game rebuilt the HUD; our snapshot is void
        if self.ws and self._ws_ready.is_set():
            try:
                self.ws.send(json.dumps({"type": P.RESYNC}))
            except Exception as e:
                logger.debug("resync send failed: %s", e)

    def _run_poll(self):
        while not self._stop.is_set():
            try:
                if not self.t.connected:
                    self.t.connect()
                if self.t.connected and not self._emu_was_connected:
                    self._resync("Emulator reconnected", wipe_baseline=False)
                    self._send_status()
                elif not self.t.connected and self._emu_was_connected:
                    self._send_status()  # emulator just went offline
                self._emu_was_connected = self.t.connected
                self._poll_once()
            except Exception as e:
                logger.debug("poll error: %s", e)
            time.sleep(self.poll_interval)

    # ── lifecycle ──────────────────────────────────────────────────────────
    def start(self):
        connected = self.t.connect()
        if not connected:
            logger.warning("No emulator/device yet — will keep retrying.")
        self._emu_was_connected = connected
        threading.Thread(target=self._run_ws, daemon=True).start()
        threading.Thread(target=self._run_poll, daemon=True).start()

    def stop(self):
        self._stop.set()
        try:
            if self.ws:
                self.ws.send(json.dumps({"type": P.BYE}))
                self.ws.close()
        except Exception:
            pass
        self.t.disconnect()

File: web/js/agent.js
/* BombosSwap - the player agent, in the browser.
 *
 * A port of agent/agent.py (HyruleAgent) without its sockets and threads:
 * snes.js owns the game link and calls tick() with the module byte every
 * loop; game.js owns the server socket and hands messages to handle(). What
 * it does, in the Python order:
 *
 *   - Pickups: every item address is read in ONE contiguous read
 *     ($7EF342-$7EF38E). First sight seeds the baseline silently; a byte
 *     equal to `expected` is our own write echoing back (accepted, not
 *     reported); otherwise each item on that byte is reported when its level
 *     went up. Several items share $7EF38C.
 *   - Save gate: only playable modules are polled. Passing through an
 *     out-of-game module (title, file select, loading, save-and-quit) means
 *     the next playable poll may be a different or reloaded save, so the
 *     baseline and expected bytes are dropped and the server is asked to
 *     re-push ownership (resync). Death, door spotlights, the mirror and
 *     boss victories are not out-of-game.
 *   - Grants and revokes: the module is read first; outside a playable
 *     module the command waits in `pending` (newest per item wins) and the
 *     next playable poll applies it before it takes the snapshot. Otherwise
 *     up to 3 tries, the result recorded in `expected`, and an `applied` ack.
 *   - Boots: the run flag ($7EF379 bit 0x04) is set every poll while the
 *     boots are ours and cleared while they are not; written only when wrong.
 *
 * Differences from the Python agent, on purpose:
 *   - A pickup that cannot be sent (server socket down) waits in `outbox`
 *     and goes out right after the next hello, so a find made during an
 *     outage is not lost to the reconnect's ownership push.
 *   - A module read that fails defers a grant or revoke too (Python only
 *     defers on a readable non-playable module and writes blind otherwise).
 *   - Every piece of game work (apply, tick) runs through one promise chain,
 *     so a grant never interleaves with a poll.
 *   - The HUD strip belongs to snes.js: it is reset there on a save load and
 *     ticked there after tick(). `say` hands a line to it.
 *
 * No DOM, no socket: node runs it against a fake WRAM.
 */
(function (root) {
  'use strict';

  var I = root.HLItems || (typeof require === 'function' ? require('./items.js') : null);
  var E = root.HLEffects || (typeof require === 'function' ? require('./effects.js') : null);

  var MODE_ADDR = I.GAME_MODE_ADDR;
  var PLAYABLE = I.PLAYABLE_MODES;
  var OUT_OF_GAME = I.OUT_OF_GAME_MODES;
  var START = I.TRACKED_START;
  var SIZE = I.TRACKED_SIZE;
  var ADDRS = I.TRACKED_ADDRS;
  var ABILITY_ADDR = E.ABILITY_ADDR;
  var RUN_ABILITY_MASK = E.RUN_ABILITY_MASK;
  var APPLY_TRIES = 3;
  var NOTIFY_MAX = 120;
  var ERROR_MAX = 200;

  var has = Object.prototype.hasOwnProperty;
  function noop() {}
  function hex2(n) { return ('0' + n.toString(16).toUpperCase()).slice(-2); }
  function errText(e) { return String(e && e.message ? e.message : e); }

  /* opts: {game: {read, write}, send(obj) -> bool (false = socket down),
            say(text), onNotify(text), onLog(text)} */
  function Agent(opts) {
    opts = opts || {};
    this.game = opts.game;
    this.fx = new E.Effects(opts.game);
    this.o = {
      send: opts.send || function () { return false; },
      say: opts.say || noop,
      onNotify: opts.onNotify || noop,
      onLog: opts.onLog || noop
    };
    this.baseline = new Map();    // addr -> last raw byte accepted as known
    this.expected = new Map();    // addr -> raw byte we just wrote (suppressed once)
    this.pending = new Map();     // item key -> [level, enable], waiting for a loaded save
    this.outOfGame = false;       // passed through title/file select/loading since the last poll
    this.bootsOwned = false;      // keeps the run flag in step every poll
    this.emu = false;             // what status reports
    this.outbox = [];             // pickups the server has not heard yet
    this._chain = Promise.resolve();
  }

  Agent.prototype._log = function (text) {
    try { this.o.onLog(text); } catch (e) { /* the page's problem */ }
  };

  /* True when the message left; false when the server socket is down. */
  Agent.prototype._send = function (obj) {
    try { return !!this.o.send(obj); } catch (e) { return false; }
  };

  /* One game job at a time: apply and tick never interleave. */
  Agent.prototype._enqueue = function (fn) {
    var self = this;
    var run = function () { return fn.call(self); };
    var next = this._chain.then(run, run);
    this._chain = next.catch(noop);
    return next;
  };

  /* server -> agent */

  /* grant | revoke | notify | reject. Resolves once a grant/revoke is done
     (or deferred). */
  Agent.prototype.handle = function (msg) {
    if (!msg || typeof msg !== 'object') return Promise.resolve();
    if (msg.type === 'grant') {
      var level = Math.trunc(Number(msg.level == null ? 1 : msg.level));
      if (!isFinite(level)) return Promise.resolve();
      return this._apply(msg.item, level, true);
    }
    if (msg.type === 'revoke') return this._apply(msg.item, 0, false);
    if (msg.type === 'notify') this._notify(msg.text);
    else if (msg.type === 'reject') this._log('Server rejected: ' + msg.reason);
    return Promise.resolve();
  };

  Agent.prototype._notify = function (text) {
    text = String(text == null ? '' : text).split(/\s+/).filter(Boolean).join(' ').slice(0, NOTIFY_MAX);
    if (!text) return;
    try { this.o.onNotify(text); } catch (e) { /* the page's problem */ }
    try { this.o.say(text); } catch (e) { /* the HUD's problem */ }
  };

  Agent.prototype._apply = function (key, level, enable) {
    return this._enqueue(function () { return this._applyNow(key, level, enable); });
  };

  /* Runs inside the chain (from _apply, or from a poll flushing `pending`). */
  Agent.prototype._applyNow = async function (key, level, enable, tries) {
    if (typeof key !== 'string' || !has.call(I.BY_KEY, key)) return;
    tries = tries || APPLY_TRIES;
    /* Only write while a save is loaded: outside a playable module the
       $7EF000 mirror is not the live save and the next file load wipes the
       write. An unreadable module byte waits too. */
    var gm = null;
    try { gm = await this.game.read(MODE_ADDR, 1); } catch (e) { gm = null; }
    var mod = gm && gm.length ? gm[0] : null;
    if (mod === null || !PLAYABLE[mod]) {
      this.pending.set(key, [level, enable]);
      this._log('Deferred ' + (enable ? 'grant' : 'revoke') + ' of ' + key + ' - '
        + (mod === null ? 'game mode unreadable' : 'no save loaded (mode 0x' + hex2(mod) + ')'));
      return;
    }
    this.pending.delete(key);     // this command supersedes any deferred one
    var last = null;
    for (var i = 0; i < tries; i++) {
      try {
        var raw = enable ? await this.fx.enable(key, level) : await this.fx.disable(key);
        this._log(enable ? 'Granted ' + key + ' (lvl ' + level + ')' : 'Revoked ' + key);
        if (key === 'boots') this.bootsOwned = enable;
        this.expected.set(I.BY_KEY[key].addr, raw);   // our write, not a pickup
        this._sendApplied(key, enable, true);
        return;
      } catch (e) {
        last = e;
      }
    }
    this._log('apply ' + key + ' failed after ' + tries + ' tries: ' + errText(last));
    this._sendApplied(key, enable, false, errText(last));
  };

  Agent.prototype._sendApplied = function (key, enable, ok, error) {
    var payload = { type: 'applied', item: key, action: enable ? 'grant' : 'revoke', ok: !!ok };
    if (error) payload.error = String(error).slice(0, ERROR_MAX);
    this._send(payload);
  };

  /* the poll */

  /* One poll, with the module byte snes.js just read. Never rejects. */
  Agent.prototype.tick = function (module) {
    return this._enqueue(function () { return this._poll(module); });
  };

  Agent.prototype._poll = async function (mod) {
    try {
      await this._pollOnce(mod);
    } catch (e) {
      this._log('poll error: ' + errText(e));
    }
  };

  Agent.prototype._pollOnce = async function (mod) {
    if (mod == null || !PLAYABLE[mod]) {
      if (mod != null && OUT_OF_GAME[mod]) this.outOfGame = true;
      return;
    }
    if (this.outOfGame) {
      this.outOfGame = false;
      /* A (re)loaded save reverts WRAM to what was last saved. Re-seed so the
         reverted bytes are not reported, and have the server re-push. */
      this.resync('Save (re)loaded', true);
    }
    await this._flushPending();

    var snap = await this.game.read(START, SIZE);
    if (!snap || snap.length < SIZE) return;

    for (var a = 0; a < ADDRS.length; a++) {
      var addr = ADDRS[a];
      var raw = snap[addr - START];
      var old = this.baseline.has(addr) ? this.baseline.get(addr) : null;
      var exp = this.expected.has(addr) ? this.expected.get(addr) : null;
      if (old === null) {
        this.baseline.set(addr, raw);             // first sight: seed
        continue;
      }
      if (exp !== null && raw === exp) {
        this.baseline.set(addr, raw);             // our own write echoing back
        this.expected.delete(addr);
        continue;
      }
      if (raw === old) continue;
      var pickups = [];
      var items = I.ITEMS_BY_ADDR[addr];
      for (var k = 0; k < items.length; k++) {
        var nl = I.discoveredLevel(items[k], raw);
        var ol = I.discoveredLevel(items[k], old);
        if (nl > ol) pickups.push([items[k].key, nl]);   // 0 -> 1, or a tier up
      }
      this.baseline.set(addr, raw);
      for (var p = 0; p < pickups.length; p++) this._sendPickup(pickups[p][0], pickups[p][1]);
    }

    var ability = START <= ABILITY_ADDR && ABILITY_ADDR < START + SIZE ? snap[ABILITY_ADDR - START] : null;
    await this._enforceBootsAbility(ability);
  };

  Agent.prototype._sendPickup = function (key, level) {
    var msg = { type: 'pickup', item: key, level: level };
    if (this._send(msg)) {
      this._log('Reported pickup: ' + key + ' lvl ' + level);
    } else {
      this.outbox.push(msg);
      this._log('Pickup of ' + key + ' held until the server is back');
    }
  };

  /* The run flag follows ownership: ALTTP can clear it on a transition
     (boots but no dash) and a lost revoke can leave it set. */
  Agent.prototype._enforceBootsAbility = async function (ability) {
    if (ability == null) {
      var data = await this.game.read(ABILITY_ADDR, 1);
      if (!data || !data.length) return;
      ability = data[0];
    }
    var have = !!(ability & RUN_ABILITY_MASK);
    if (this.bootsOwned && !have) {
      await this.game.write(ABILITY_ADDR, new Uint8Array([(ability | RUN_ABILITY_MASK) & 0xFF]));
    } else if (!this.bootsOwned && have) {
      await this.game.write(ABILITY_ADDR, new Uint8Array([ability & ~RUN_ABILITY_MASK & 0xFF]));
    }
  };

  Agent.prototype._flushPending = async function () {
    var list = Array.from(this.pending.entries());
    this.pending.clear();
    for (var i = 0; i < list.length; i++) {
      await this._applyNow(list[i][0], list[i][1][0], list[i][1][1]);
    }
  };

  /* link events */

  /* Ask the server to re-push ownership. wipeBaseline (default true) re-seeds
     detection from the game without reporting it: right for a save reload,
     wrong for a link blip on the same save, where the old baseline is still
     valid and a pickup made during the blip must still be reported. */
  Agent.prototype.resync = function (reason, wipeBaseline) {
    var wipe = wipeBaseline !== false;
    this._log((reason || 'Resync') + ' - '
      + (wipe ? 're-seeding state and requesting resync' : 'requesting resync') + '.');
    if (wipe) {
      this.baseline.clear();
      this.expected.clear();
    }
    this._send({ type: 'resync' });
  };

  /* The agent socket just said hello. */
  Agent.prototype.onServerOpen = function () {
    this._send({ type: 'status', emu: !!this.emu });
    var box = this.outbox;
    this.outbox = [];
    for (var i = 0; i < box.length; i++) {
      if (!this._send(box[i])) {
        this.outbox = box.slice(i).concat(this.outbox);
        return;
      }
      this._log('Reported pickup: ' + box[i].item + ' lvl ' + box[i].level + ' (held)');
    }
  };

  /* first = the first attach of this page: the hello's ownership push is
     still to come, so there is nothing to resync. */
  Agent.prototype.onGameAttached = function (first) {
    this.emu = true;
    if (!first) this.resync('Game link back', false);
    this._send({ type: 'status', emu: true });
  };

  Agent.prototype.onGameLost = function () {
    this.emu = false;
    this._send({ type: 'status', emu: false });
  };

  Agent.prototype.setEmu = function (on) {
    this.emu = !!on;
  };

  var HLAgent = { Agent: Agent, APPLY_TRIES: APPLY_TRIES };
  root.HLAgent = HLAgent;
  if (typeof module === 'object' && module.exports) module.exports = HLAgent;
})(typeof window !== 'undefined' ? window : globalThis);
